### DISCUSSION TITLE
Автоматическое переключение на созданный плейлист

### DISCUSSION THREAD

**Discussion by Bedmit** [Chimildic/goofy#125] [Помощь с алгоритмом] [answered]
В связи с появившейся возможностью использовать Таскер, модернизирую свой старый скрипт. По встряхиванию телефона при каком-то включенном плейлисте, он доложен создать плейлист, сдвинув не прослушанные треки вверх и заменив остальные на похожие.
```
function Shake() {
let playlistId = Helper.parseId(Player.getPlayback().context.uri);
let Fresh = Source.getPlaylistTracks('', playlistId);
  let FreshOld = Selector.sliceCopy(Fresh);

  let LikedTracks = Cache.read('LikedTracks.json');
    Filter.replaceWithSimilar({
    origin: Fresh,
    replace: [RecentTracks.get(5000), LikedTracks],
    });
  Filter.matchExcept(Fresh, 'mix|club|radio|edit|live|karaoke');
  Filter.dedupTracks(Fresh);
  Filter.removeUnavailable(Fresh);

  Filter.removeTracks(FreshOld, Fresh, true);
  Filter.removeTracks(Fresh, FreshOld);
  Combiner.push(FreshOld, Fresh);

  let playlist = Playlist.getById(playlistId);

  Playlist.saveWithReplace({
    name: playlist.name + ' *',
    description: 'По мотивам ' + playlist.name,
    public: false,
    sourceCover: playlist.images[0].url,
    tracks: FreshOld,  
  });  
}
```
Хочу еще дополнить его переходом на этот созданный плейлист. Понимаю, что это делается с помощью 
```
Player.resume({
    context_uri: `spotify:playlist:${playlistId}`,
});
```
Но как получить ID созданного плейлиста? Или можно осуществить его включение как-то по другому?

И еще нюанс. 
Этот сценарий используется для чужих, не редактируемых плейлистов. По ним создается измененный дубликат. Для удобства он помечается звездочкой. Если этот дубликат уже есть, он просто перезаписывается. А если при прослушивании этого дубликата запустить скрипт, создается еще один дубликат дубликата, с двумя звездочками.
Как сделать, чтобы в этом случае просто перезаписывался тот который включен? Смог сделать это возможным, только отключив добавление звездочки в название. Но тогда не так удобно отличать дубликат от оригинала.
Если это сложно, то не обязательно. Сгодится и вариант без звездочки.

**Comment by Chimildic** [comment, marked as answer]
> как получить ID созданного плейлиста?

Обновил библиотеку. Теперь функции `Playlist.save*` вернут _id_ плейлиста, в который добавлялись треки. При любом сценарии: существующий плейлист, найденный по имени или только что созданный. 

```js
let remotePlaylistId = Playlist.saveWithReplace({
   // ...
}); 
```

При каком условии должен случится переход по `Player.resume`?

> Как сделать, чтобы в этом случае просто перезаписывался тот который включен?

_Маркером_ плейлиста будет не то, что он включен, а само имя. Достаточно сделать условие с проверкой. 

```js
// Если звездочка всегда последний символ, используем slice с аргументом -1
// Когда она есть, название не изменится, т.е. обновится существующий плейлист
Playlist.saveWithReplace({
  name: playlist.name.slice(-1) == '*' ? playlist.name : `${playlist.name} *`,
  // ...
}); 

// Если звездочка первая, просто нулевой индекс
Playlist.saveWithReplace({
  name: playlist.name[0] == '*' ? playlist.name : `${playlist.name} *`,
  // ...
}); 
```

**Reply by Bedmit** [reply]
> При каком условии должен случится переход по `Player.resume`?

Переход безусловный. Включили плейлист -> встряхнули телефон (ну или какой там триггер) -> сгенертровался и включился дубликат с замененными треками
Конечный вариант со всем справляется 
```js
function Shake() {
  const PLAYLIST_ID = '74uSA9FulwvWqVSi7mxWyQ';
  let playback = Player.getPlayback();
  if (playback.isEmpty()
    || playback.context.type != 'playlist'
    || playback.context.uri.includes(PLAYLIST_ID)
    || playback.repeat_state == 'track'
    || playback.shuffle_state) {
    return;
  }

  let playlistId = Helper.parseId(Player.getPlayback().context.uri);
  let Fresh = Source.getPlaylistTracks('', playlistId);
  let FreshOld = Selector.sliceCopy(Fresh);

  let LikedTracks = Cache.read('LikedTracks.json');
  Filter.replaceWithSimilar({
    origin: Fresh,
    replace: [RecentTracks.get(5000), LikedTracks],
  });
  Filter.matchExcept(Fresh, 'mix|club|radio|edit|live|karaoke');
  Filter.dedupTracks(Fresh);
  Filter.removeUnavailable(Fresh);

  Filter.removeTracks(FreshOld, Fresh, true);
  Filter.removeTracks(Fresh, FreshOld);
  Combiner.push(FreshOld, Fresh);

  let playlist = Playlist.getById(playlistId);

  let remotePlaylistId = Playlist.saveWithReplace({
    name: playlist.name.slice(-1) == '*' ? playlist.name : `${playlist.name} *`,
    description: 'По мотивам ' + playlist.name,
    public: false,
    sourceCover: playlist.images[0].url,
    tracks: FreshOld,
  });

  Player.resume({
    context_uri: `spotify:playlist:${remotePlaylistId}`,
  });
}
```

После запуска скрипта естественно проходит какое-то время, но это буквально секунды. Забавно, что сам плейлист еще даже не успевает появиться в фонотеке, а уже запускается и играет. 
Я как всегда доволен, жду когда меня это перестанет удивлять ))

**Reply by Chimildic** [reply]
Заметил, что код не отформатирован. В редакторе _правая кнопка мыши > форматировать документ_ или комбинация <kbd>Shift</kbd><kbd>Alt</kbd><kbd>F</kbd>. Сделает более ровную структуру по отступам, легче читается.

Кстати, чтобы включить подсветку кода здесь, на форуме
```js
let remotePlaylistId = Playlist.saveWithReplace({
   // ...
}); 
```
Нужна такая структура - это символ с кнопки <kbd>ё</kbd> по три раза в начале и конце, плюс приписка языка `js`

![image](https://user-images.githubusercontent.com/40138097/133133441-e471e804-bf4f-4a3b-ab3d-6428ba7e6fe8.png)

**Reply by Bedmit** [reply]
Спасибо. Все причесал. А главное и у себя порядок теперь

**Reply by Chimildic** [reply]
`js` только в начале, в конце не нужно

**Comment by Bedmit** [comment]
Сделал себе еще подобную штуку по треку. Тоже все ок. Но один нюанс. 

Сгенерированный плейлист какое-то время добирается до клиента и стартует его воспроизведение. Буквально несколько секунд. И это время плейбэк выгружается из всех клиентов и естественно ни чего не воспроизводится. Выглядит не очень.
Но увидел в описании `resume` такую штуку - `Utilities.sleep(5000);`
Как я понял, это пауза между выполнениями строк кода в миллисекундах. Оказалось это как раз то что нужно, чтобы не прерывалось воспроизведение.
Может есть смысл в инструкции тоже  это как-то отметить?

```js
function RadioTask() {
  track = Player.getPlayback().item;
  let seed = [track];
  let radioTracks = Source.craftTracks(seed);
  let cacheLiked = Cache.read('LikedTracks.json');
  let historyTracks = RecentTracks.get(5000);

  Order.sort(radioTracks, 'meta.popularity', 'desc');

  Filter.matchExcept(radioTracks, 'mix|club|radio|edit|live|karaoke');
  Filter.removeTracks(radioTracks, cacheLiked);
  Filter.removeTracks(radioTracks, historyTracks);
  Filter.rangeDateRel(historyTracks);
  Filter.removeArtists(radioTracks, historyTracks);
  Filter.removeUnavailable(radioTracks);
  Filter.dedupArtists(radioTracks);

  let tracks = Combiner.push(seed, radioTracks);

  let remotePlaylistId = Playlist.saveWithReplace({
    name: tracks[0].name + ' &',
    tracks: tracks,
    description: Playlist.getDescription(tracks),
    sourceCover: tracks[0].album.images[0].url,
    public: false,
  });

  Utilities.sleep(20000);
  Player.resume({
    context_uri: `spotify:playlist:${remotePlaylistId}`,
  });
}

```

**Reply by Chimildic** [reply]
> в инструкции тоже это как-то отметить?

В инструкции чего?

**Reply by Bedmit** [reply]
Goofy.
Например прямо там же в примере, где она упоминается https://chimildic.github.io/goofy/#/func?id=resume
Или, если это может понадобится в любых функциях, например здесь https://chimildic.github.io/goofy/#/best-practices?id=%d0%9f%d0%be%d0%bb%d0%b5%d0%b7%d0%bd%d1%8b%d0%b9-%d0%be%d0%bf%d1%8b%d1%82
Не настаиваю конечно. Я как и в прошлом подобном случае, не очень уверен, что столько затраченных сил на все эти описания оправданы.
